File: Q21/Q21.java
import java.util.PriorityQueue;
import java.util.Arrays;

public class Q21{
    public ListNode mergeTwoLists(ListNode l1, ListNode l2){
	// Data stuctures and variables
	PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
	ListNode answer, temp;
	
	// Unwrap the list nodes
	while(l1 != null){
	    pq.add(l1.val);
	    l1 = l1.next;
	}
	while(l2 != null){
	    pq.add(l2.val);
	    l2 = l2.next;
	}
	
	// Build new sorted list
	if(pq.peek() != null){
	    answer = new ListNode(pq.poll());
	    temp = answer;
	    while(pq.peek() != null){
		temp.next = new ListNode(pq.poll());
		temp = temp.next;
	    }
	}else{
	    answer = null;
	}
	
	// Return result
	return answer;
    }
}


/*
Runtime: 2 ms, faster than 21.50% of Java online submissions for Merge Two Sorted Lists.
Memory Usage: 38.4 MB, less than 19.53% of Java online submissions for Merge Two Sorted Lists.
*/

File: Q21/Q21.py
from ListNode import ListNode

class Q21:
    def mergeTwoLists(self, l1, l2):
        # unwrap the nodes to an array
        combined = []
        
        while(l1 != None):
            combined.append(l1.val)
            l1 = l1.next

        while(l2 != None):
            combined.append(l2.val)
            l2 = l2.next

        # check length and sort
        if combined == []:
            return None

        combined.sort()

        # Build new list and return
        answer = ListNode(combined[0])
        temp = answer
        combined.pop(0)
        for x in combined:
            temp.next = ListNode(x)
            temp = temp.next

        return answer
    
"""
Runtime: 48 ms, faster than 26.42% of Python3 online submissions for Merge Two Sorted Lists.
Memory Usage: 13.7 MB, less than 6.61% of Python3 online submissions for Merge Two Sorted Lists.
"""
